thorlcr/activities/hashdistrib/thhashdistribslave.cpp

```diff
@@ -2924,12 +3060,29 @@ class HashDedupSlaveActivityBase : public CSlaveActivity
                 lastEog = false;
                 if (bucketHandler->addRow(row)) // true if new, i.e. non-duplicate (does not take ownership)
                 {
-                    dataLinkIncrement();
-                    return row.getClear();
+                    if (!keepBest)
+                    {
+                        dataLinkIncrement();
+                        return row.getClear();
+                    }
                 }
             }
             else
             {
+                if (keepBest && !keepBestRowsReadyToReturn)
+                {
+                    bucketHandler->clearCallbacks();
```

Comment:
What if it has spilt? 
I don't the following code is right if so. i.e. it's going to reading and returning partially spilt buckets.

Or some buckets may be spilt, others not.
You need to output all the unspilt bucket only at this state, but let the process of descending into next bucket handlers continue. See getNextBucketHandler() and loop changing distInput row stream.

For testing purposes I think you should add something that forces spilling at this point via a testing option, something like:
```
    unsigned testSpillTimes = getOptInt("testHashDedupSpillTimes");
```

.. in constructor

and :
```
    if (testSpillTimes)
    {
        bucketHandler->spillBucket(false);
        testSpillTimes--;
    }
```

just after the clearCallbacks();

That will force a spill event and another cycle of buckets.